brownie/network/multicall.py

```diff
@@ -0,0 +1,157 @@
+
+
+class LazyResult(Proxy):
+    """A proxy object to be updated with the result of a multicall."""
+
+    def __repr__(self) -> str:
+        return repr(self.__wrapped__)
+
+
+class Multicall:
+    """Context manager for batching multiple calls to constant contract functions."""
+
+    def __init__(self) -> None:
+        self.address = None
+        self.block_number = None
+        self._contract = None
+        self._pending_calls: List[Call] = []
+
+        setattr(ContractCall, "__original_call_code", ContractCall.__call__.__code__)
+        setattr(ContractCall, "__proxy_call_code", self._proxy_call.__code__)
+        setattr(ContractCall, "__multicall", defaultdict(lambda: None))
+        ContractCall.__call__.__code__ = self._proxy_call.__code__
+
+    def __call__(
+        self, address: Optional[str] = None, block_identifier: Union[str, bytes, int, None] = None
+    ) -> "Multicall":
+        self.address = address  # type: ignore
+        self.block_number = block_identifier  # type: ignore
+        return self
+
+    def _flush(self, future_result: Result = None) -> Any:
+        if not self._pending_calls:
+            # either all calls have already been made
+            # or this result has already been retrieved
+            return future_result
+        ContractCall.__call__.__code__ = getattr(ContractCall, "__original_call_code")
+        results = self._contract.tryAggregate(  # type: ignore
+            False,
+            [_call.calldata for _call in self._pending_calls],
+            block_identifier=self.block_number,
+        )
+        ContractCall.__call__.__code__ = getattr(ContractCall, "__proxy_call_code")
+        for _call, result in zip(self._pending_calls, results):
+            _call.__wrapped__ = _call.decoder(result[1]) if result[0] else None  # type: ignore
+        self._pending_calls = []  # empty the pending calls
+        return future_result
+
+    def flush(self) -> Any:
+        """Flush the pending queue of calls, retrieving all the results."""
+        return self._flush()
+
+    def _call_contract(self, call: ContractCall, *args: Tuple, **kwargs: Dict[str, Any]) -> Proxy:
+        """Add a call to the buffer of calls to be made"""
+        calldata = (call._address, call.encode_input(*args, **kwargs))  # type: ignore
+        call_obj = Call(calldata, call.decode_output)  # type: ignore
+        # future result
+        result = Result(call_obj)
+        self._pending_calls.append(result)
+
+        return LazyResult(lambda: self._flush(result))
+
+    @staticmethod
+    def _proxy_call(*args: Tuple, **kwargs: Dict[str, Any]) -> Any:
+        """Proxy code which substitutes `ContractCall.__call__`
+
+        This makes constant contract calls look more like transactions since we require
+        users to specify a dictionary as the last argument with the from field
+        being the multicall2 instance being used."""
+        from threading import get_ident
```

Comment:
Is this import statement necessary?

Reply:
Ehh it is gross, what I can do is move the import over to the contract.py file then we won't have to import in the function